### PR TITLE
Soil Moisture Sensor Support

### PR DESCRIPTION
**Pull request by alistair23**
### Pull Request Overview

This adds support for reading the soil moisture (from the new moisture HIL). This also includes support for the soil moisture sensor from the [Chirp project](https://www.tindie.com/products/miceuz/i2c-soil-moisture-sensor/).

This is all based on the existing humidity sensor, as they are both similar concepts, but slightly different so I think they need a new sensor type.

This also unfortunately adds another Cargo feature to the Apollo3 board, which is a bit clunky. But there isn't really a better way to allow adding support for sensors

### Testing Strategy

The unit tests in the PR

### Documentation Updated

- [X] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [X] Ran `make prepush`.

### CODE REVIEW

**Comment by alevy** [comment]
The HIL looks fine to me. @bradjc you have a moisture sensor where, I believe, you're just passing the ADC up to userspace? Would this HIL work for your case as well?

File: capsules/extra/src/moisture.rs
Comment (bradjc) [inline capsules/extra/src/moisture.rs:106]: If this fails it needs to unset `busy`.

File: capsules/extra/src/moisture.rs
Comment (bradjc) [inline capsules/extra/src/moisture.rs:132]: It would be good to use the first argument as an ReturnCode.

File: capsules/extra/src/moisture.rs
Comment (bradjc) [inline capsules/extra/src/moisture.rs:20]:
```suggestion
```

**Review by bradjc — CHANGES_REQUESTED** [review]

**Comment by alistair23** [comment]
Whoops, I forgot to add a paragraph in the PR description.

The moisture HIL and userspace interface is an exact copy of the humidity API. The idea is that humidity and moisture are similar so they should have the same interface. This should make it easy for HIL users and applications to switch between the two. It's also simple to implement.

This comes at a cost of inheriting the issues of the original humidity HIL and API, most notably a lack of error reporting.

I'm happy to improve the moisture HIL and syscall API, breaking compatibility with the existing humidity users if that's the preferred approach. @bradjc seems to think it's worth fixing the issues for the moisture API. I'll give this a few days to see if anyone else feels strongly either way.

**Comment by alevy** [comment]
@alistair23 I think @bradjc's comments on error reporting are split in two (two specific comments in his otherwise comprehensive review):

1. Does the HIL bubble errors up?
2. Does the system call driver bubble errors up?

It seems pretty clear to me the HIL should bubble errors up. The humidity HIL is just old and not up to the times. Note, for example, that the temperature HIL and others _do_ have errors. It probably just makes sense to update the humidity HIL (separately).

The humidity system call driver can't be changed except in backwards compatible ways until 2.0. This is true for other sensor drivers, like the temperature stack, that do bubble errors up from the HIL but ignore errors in the system call driver.

I'm personally fine with either making the pre-2.0 system call driver for moisture the aligned with other sensor drivers _or_ improving over them in the moisture driver by bubbling an error up in the first argument.

File: capsules/extra/src/chirp_i2c_moisture.rs
Comment (alistair23) [reply capsules/extra/src/chirp_i2c_moisture.rs:71]: We don't have to run this. My worry is that if we don't check the version before starting that there might be some breaking change in the future that we won't notice (or there has been a breaking change in the past)

**Comment by alistair23** [comment]
Updated! I just went with updating to provide errors to the HIL and the userspace API as that seems to be the consensus

**Comment by alistair23** [comment]
Ping!

**Comment by alistair23** [comment]
Ping!

**Comment by alistair23** [comment]
Ping!

File: boards/components/src/chirp_i2c_moisture.rs
Comment (bradjc) [inline boards/components/src/chirp_i2c_moisture.rs:5]:
```suggestion
//! Components for the Chirp I2C Moisture Sensor.
```

File: boards/apollo3/lora_things_plus/Makefile
Comment (bradjc) [inline boards/apollo3/lora_things_plus/Makefile:63]: I'm trying to figure this out for the other PR, but why do we need to copy around linker files to run tests?

**Review by bradjc — COMMENTED** [review]
Just a couple small things.

File: boards/apollo3/lora_things_plus/Makefile
Comment (alistair23) [reply boards/apollo3/lora_things_plus/Makefile]: It seems like we don't need it any more, I have removed them

**Review by alevy — APPROVED** [review]

**Comment by alistair23** [comment]
Rebased, this is ready to go

File: boards/apollo3/lora_things_plus/src/main.rs
Comment (lschuermann) [inline boards/apollo3/lora_things_plus/src/main.rs:120]: So I know that this is following the established pattern, but is actually really problematic in practice now that use of static muts is a warning (and will be promoted to a hard error in the next Rust edition). I don't know whether it's worth blocking this PR any further on, but we will have to go through all of these tests and remove those static usages eventually. Right now this only passes our CI because we don't check for warnings when we run tests (which is bad). I devised a fix for this that I should upstream at some point, but it'll obviously make all of these tests fail.

I think we'll instead want reusable setup functions (like this board has anyways),, that a `#[test]` function then calls to instantiate the peripherals and pass `&'static mut` references to the functions that actually execute the tests. I had a much better grasp on this a while ago, I'll see whether I can dig this back up.

File: kernel/src/hil/sensors.rs
Comment (lschuermann) [inline kernel/src/hil/sensors.rs:49]: The HIL is simple enough, but what are the error types that this can return? I don't know whether there's anything special about the moisture sensor, so perhaps we can introduce (if it doesn't already exist) and refer to a standard list of error variants that all of these types of HILs can return (e.g., `ErrorCode::OFF`, which has pretty clear semantics across all sensors).

**Review by lschuermann — COMMENTED** [review]
Reviewed just the HIL and tests right now.

File: boards/apollo3/lora_things_plus/src/main.rs
Comment (alistair23) [reply boards/apollo3/lora_things_plus/src/main.rs:120]: I think that all of the tests in a range of boards will need to be updated to not use static muts and this PR shouldn't block on that. We can merge this and then still update all of the static muts in the future

File: kernel/src/hil/sensors.rs
Comment (alistair23) [reply kernel/src/hil/sensors.rs:57]: I added documentation on the possible errors

File: boards/apollo3/redboard_artemis_atp/Makefile
Comment (bradjc) [inline boards/apollo3/redboard_artemis_atp/Makefile:47]: Do we need these two makefile changes?

**Review by bradjc — DISMISSED** [review]
Looked at capsules and not the tests.

File: boards/apollo3/lora_things_plus/src/main.rs
Comment (lschuermann) [reply boards/apollo3/lora_things_plus/src/main.rs:120]: We discussed this on the core call today. I think that you make a fair argument by saying that this follows an established paradigm.

Ultimately, this is a very unfortunate situation. A while back we accidentally started permitting warnings in the CI tests. Then, by updating the Rust toolchain, we introduced these `static mut` warnings but did not notice how prevalent this pattern was in the integration tests. Now, we're faced with the prospect of these warnings **becoming hard errors in the 2024 Rust edition**, which will be released imminently.

I strongly believe that it's not solely on you to fix these errors. Furthermore, for these tests, we don't want to simply fix them by using raw pointer dereferences -- the hack that we used to get rid of these warnings on master was really just that, a hack, and I'd rather not have more of those in the codebase stick around forever. Instead, we'll want to refactor those tests to instead pass such instances as parameters and avoid going via global variables entirely.

However, I don't think that adding _more_ of these tests is a good idea either. To avoid blocking this PR, can we suggest to split out the tests (which already confirm that this code works!) into a different PR? This would avoid introducing more mutable statics into our codebase, and we can also use that PR to see how we can convert these existing tests into ones that don't use mutable statics.

**Review by lschuermann — CHANGES_REQUESTED** [review]
See https://github.com/tock/tock/pull/4144#discussion_r1817287509

File: boards/apollo3/redboard_artemis_atp/Makefile
Comment (alistair23) [reply boards/apollo3/redboard_artemis_atp/Makefile:47]: It was part of a larger cleanup, but a similar change got merged separately. So this is now just what's left.

It does make it easier to read though, before it was a really long line

File: boards/apollo3/lora_things_plus/src/main.rs
Comment (alistair23) [reply boards/apollo3/lora_things_plus/src/main.rs]: Done!

**Review by alevy — DISMISSED** [review]

File: capsules/extra/src/chirp_i2c_moisture.rs
Comment (lschuermann) [reply capsules/extra/src/chirp_i2c_moisture.rs:72]: Can we call this `initialize` please? Regardless of whether this is required or not, if you're _supposed_ to call it before using the chip, then I'd view that as part of its initialization routine, and it should follow the nomenclature that we use elsewhere and also documented (albeit for HILs) in TRD3.

**Review by lschuermann — DISMISSED** [review]
Seems good other than my one comment.

**Review by lschuermann — APPROVED** [review]

**Comment by lschuermann** [comment]
This is already last-call, let's wait another few hours before merge.

**Review by alevy — APPROVED** [review]

**Pull request merged**
